File: Java/Fibonacci.java
/* ==================================================
 * Fibonacci
 * Shikuan Huang
 * 
 * Basic recursive implementation
 * Dynamic programming implementation
 * Tail recursive implementation
 * ==================================================
 */

import java.math.BigInteger;
import java.util.*;

public class Fibonacci {
  public static void main(String[] args) {
	Scanner input = new Scanner(System.in);
	System.out.println("Fibonacci of? ");
	long userInput = input.nextLong();
	BigInteger converted = new BigInteger(String.valueOf(userInput));
	System.out.println("Fib(" + converted + ") = " + iFib(converted));
	input.close();
  }
  
  // Basic recursive implementation
  static BigInteger rFib(BigInteger arg) {
    if (arg.equals(BigInteger.ZERO) || arg.equals(BigInteger.ONE)) {
      return arg;
    }
    
    BigInteger TWO = new BigInteger(String.valueOf(2));
    return rFib(arg.subtract(BigInteger.ONE)).add(rFib(arg.subtract(TWO)));
  }
  
  // Dynamic programming implementation
  static BigInteger dFib(BigInteger arg) {
	BigInteger[] answers = new BigInteger[arg.intValue() + 1];
	Arrays.fill(answers, new BigInteger(String.valueOf(-1)));
    return dFib(arg, answers);
  }
  
  static BigInteger dFib(BigInteger arg, BigInteger[] answers) {
	BigInteger NegOne = new BigInteger(String.valueOf(-1));
    if (!answers[arg.intValue()].equals(NegOne)) {
      return answers[arg.intValue()];
    }
    
    if (arg.equals(BigInteger.ZERO) || arg.equals(BigInteger.ONE)) {
      answers[arg.intValue()] = arg;
      return arg;
    }
    
    BigInteger TWO = new BigInteger(String.valueOf(2));
    answers[arg.intValue()] = dFib(arg.subtract(BigInteger.ONE), answers)
        .add(dFib(arg.subtract(TWO), answers));
    return answers[arg.intValue()];
  }
  
  // Tail recursive implementation
  static BigInteger tFib(BigInteger arg) {
	if (arg.equals(BigInteger.ZERO)) {
	  return arg;
	}
	
	return tFib(arg.subtract(BigInteger.ONE), BigInteger.ZERO, BigInteger.ONE);
  }
  
  static BigInteger tFib(BigInteger arg, BigInteger first, BigInteger second) {
	if (arg.equals(BigInteger.ZERO)) {
	  return second;
	}
	
	return tFib(arg.subtract(BigInteger.ONE), second, first.add(second));
  }
  
  // Iterative implementation
  static BigInteger iFib(BigInteger arg) {
	if (arg.equals(BigInteger.ZERO)) {
	  return arg;
	}
    BigInteger first = BigInteger.ZERO;
    BigInteger second = BigInteger.ONE;
    BigInteger temp = BigInteger.ZERO;
    
    while (!arg.equals(BigInteger.ONE)) {
      temp = first.add(second);
      first = second;
      second = temp;
      arg = arg.subtract(BigInteger.ONE);
    }
    
    return temp;
  }
}

File: CPP/Fibonacci.cpp
/* ==================================================
Fibonacci
Shikuan Huang
================================================== */

#include <iostream>
using namespace std;

// Basic recursive implemenation
long long rFibonacci(int number) {
	if (number < 0) {
		throw number;
	}
		
	if (number == 0 || number == 1) {
		return number;
	}
	else {
		return rFibonacci(number - 1) + rFibonacci(number - 2);
	}
}

// Iterative implementation
long long iFibonacci(int number) {
	if (number < 0) {
		throw number;
	}
		
	if (number == 0) {
		return 0;
	}
		
	long long first = 0;
	long long second = 1;
	long long next = 0;
	
	for (int i = 1; i < number; i++) {
		next = first + second;
		first = second;
		second = next;
	}
	
	return second;
}

// Helper function that performs the actual computation
long long dFib(int number, long long * answers) {	
	if (number == 0 || number == 1) {
		return number;
	}
	
	if (answers[number - 1] != -1) {
		return answers[number - 1];
	}
	
	answers[number - 1] = dFib(number - 1, answers) + dFib(number - 2, answers);
	return answers[number - 1];
}

// Dynamic programming implementation
long long dFibonacci(int number) {
	if (number < 0) {
		throw number;
	}
	
	long long * answers = new long long[number];
	for (int i = 0; i < number; i++) {
		answers[i] = -1;
	}
		
	long long answer = dFib(number, answers);
	delete[] answers;
	return answer;
}

int main() {
	int input;
	bool end = false;
	while (!end) {
		cout << "factorial for? ";
		cin >> input;
		if (input < 0) {
			cout << "INVALID INPUT\n\n";
		}
		else {
			cout << "fibonacci(" << input << ") = " << dFibonacci(input) << endl << endl;
		}
	}
}
